java/sage/nio/RemoteFileChannel.java

```diff
@@ -0,0 +1,703 @@
+      // don't see an exception until we try to read.
+      return readMore(dst);
+    }
+    catch (IOException e)
+    {
+      if (retry)
+      {
+        reconnect();
+        return startRead(position, count, dst, false);
+      }
+      else
+      {
+        throw e;
+      }
+    }
+  }
+
+  private int readMore(ByteBuffer dst) throws IOException
+  {
+    if (!dst.hasRemaining())
+      return 0;
+
+    int returnValue = -1;
+
+    while (dst.hasRemaining())
+    {
+      int readBytes = socket.read(dst);
+
+      if (readBytes == -1)
+        break;
+
+      if (returnValue == -1)
+        returnValue = 0;
+
+      returnValue += readBytes;
+    }
+
+    return returnValue;
+  }
+
+  // You must synchronize startWrite and writeMore in the method using these so no other
+  // communication can be attempted while in use.
+  private int startWrite(long position, long count, ByteBuffer src, boolean retry) throws IOException
+  {
+    try
+    {
+      commBuf.clear();
+      commBuf.put(("WRITE " + position + " " + count + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      commBuf.flip();
+      socket.write(commBuf);
+      return writeMore(src);
+    }
+    catch (IOException e)
+    {
+      if (retry)
+      {
+        reconnect();
+        return startWrite(position, count, src, false);
+      }
+      else
+      {
+        throw e;
+      }
+    }
+  }
+
+  private int writeMore(ByteBuffer src) throws IOException
+  {
+    int returnValue = 0;
+
+    while (src.hasRemaining())
+    {
+      returnValue += socket.write(src);
+    }
+
+    return returnValue;
+  }
+
+  ByteBuffer singleByte;
+  @Override
+  public int readUnsignedByte() throws IOException
+  {
+    if (singleByte == null)
+    {
+      singleByte = ByteBuffer.allocate(1);
+    }
+    singleByte.clear();
+    int bytes = read(singleByte);
+    singleByte.flip();
+
+    if (bytes == -1)
+      throw new java.io.EOFException();
+
+    return singleByte.get() & 0xFF;
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException
+  {
+    int returnValue = read(dst, remoteOffset);
+
+    if (returnValue == -1)
+      return returnValue;
+
+    remoteOffset += returnValue;
+    return returnValue;
+  }
+
+  @Override
+  public synchronized long read(ByteBuffer[] dsts, int offset, int length) throws IOException
+  {
+    if (length == 0)
+      return 0;
+
+    long returnValue = 0;
+    long readBytes;
+    long totalLength = 0;
+
+    int remains;
+    boolean bufferAvailable = false;
+    int limit = offset + length;
+    for (int i = offset; i < limit; i++)
+    {
+      remains = dsts[i].remaining();
+
+      // Since we are iterating through the buffers either way, we might as well find the first one
+      // that actually has space available.
+      if (remains > 0)
+      {
+        bufferAvailable = true;
+        totalLength += remains;
+      }
+      else if (!bufferAvailable)
+      {
+        offset += 1;
```

Comment:
Actually, I think you're wrong. I traced every step of this this through the debugger and it does exactly what it should be doing. When the first buffer has space and the next few buffers don't have anything, the offset is still 0. When the first buffer has nothing and the second buffer has space, the offset is returned as 1.

I  looked the problem so hard I went from not seeing it to actually creating a new problem in my last commit.